internal/controller/pipelinerollout_controller.go

```diff
@@ -424,74 +424,102 @@ func (r *PipelineRolloutReconciler) processExistingPipeline(ctx context.Context,
 	}
 	return nil
 }
+
```

Comment:
the diff below makes it hard to see what changed. Basically, I separated out a bunch of the code from `processExistingPipelineWithoutDataLoss()` and moved it into a new function called `shouldBePaused()` for readability.